tests/fpgadataflow/test_fpgadataflow_thresholding_binary_search.py

```diff
@@ -0,0 +1,495 @@
+        model = model.transform(SetExecMode("cppsim"))
+    except Exception as e:
+        if (
+            str(e)
+            != "Custom op_type Thresholding_Binary_Search is currently not supported."
+        ):
+            raise
+
+
+# Test brief: Prove that memory mode 'const' is not supported for this layer type
+@pytest.mark.fpgadataflow
+@pytest.mark.vivado
+def test_fpgadataflow_thresholding_binary_search_const_mem_mode():
+    input_data_type = DataType["INT16"]
+    activation = DataType["INT4"]
+    fold = -1
+    num_input_channels = 16
+    mem_mode = "const"
+
+    pe = generate_pe_value(fold, num_input_channels)
+    num_input_vecs = [1, 2, 2]
+    output_data_type = activation
+    activation_bias = output_data_type.min()
+
+    # Generate random thresholds and sort in ascending order
+    num_steps = activation.get_num_possible_values() - 1
+    thresholds = generate_random_threshold_values(
+        input_data_type, num_input_channels, num_steps
+    )
+
+    # Generate model from input parameters to the test
+    model = make_single_thresholding_binary_search_modelwrapper(
+        thresholds,
+        pe,
+        input_data_type,
+        output_data_type,
+        activation_bias,
+        mem_mode,
+        num_input_vecs,
+    )
+
+    # Prove that 'const' memory mode is not supported for this class
+    # 'const' memory mode is not supported for this class, catch the specific exception
+    # thrown by FINN. Exception: ('Unrecognized memory mode for this node:', 'const')
+    try:
+        model = model.transform(InsertFIFO(True))
+        model = model.transform(GiveUniqueNodeNames())
+        model = model.transform(PrepareIP(test_fpga_part, target_clk_ns))
+        model = model.transform(HLSSynthIP())
+        model = model.transform(CreateStitchedIP(test_fpga_part, target_clk_ns))
+    except Exception as e:
+        if str(e) != "Unrecognized memory mode for this node: {}".format(mem_mode):
+            raise
+        # Caught the expected exception, leave the test early
+        return
+
+
+# Test brief: Test that PrepareRTLSim() runs successfully. This function is not
+# tested in test_fpgadataflow_thresholding_binary_search()
+@pytest.mark.fpgadataflow
+@pytest.mark.vivado
+def test_fpgadataflow_thresholding_binary_search_prepare_rtlsim():
+    input_data_type = DataType["INT16"]
+    act = DataType["INT4"]
+    fold = -1
+    num_input_channels = 16
+    mem_mode = "decoupled"
+
+    # Handle inputs to the test
+    pe = generate_pe_value(fold, num_input_channels)
+    num_steps = act.get_num_possible_values() - 1
+
+    # Generate random, non-decreasing thresholds
+    thresholds = generate_random_threshold_values(
+        input_data_type, num_input_channels, num_steps
+    )
+    # make the vivado_hls threshold bug appear (incorrect rtlsim result when first
+    # threshold of first channel is zero, while using BIPOLAR output)
+    if act == DataType["BIPOLAR"]:
+        thresholds[0][0] = 0
```

Comment:
I don't think this is relevant for the RTL thresholding, this was a workaround to trigger an HLS bug

Reply:
removed in https://github.com/Xilinx/finn/pull/687/commits/f79b9ec3e19d83d6469e6e563422fbba70f7a87a